tests/contrib.print_proxy.mapfish_print/test_mapfish_print.py

```diff
@@ -261,6 +261,49 @@ def test_get_sorted_legal_provisions(DummyRenderInfo):
                 "http://www.bav.admin.ch/themen/verkehrspolitik/00709/index.html",
             "TextAtWeb": [{"URL": "https://oereb-gr-preview.000.ch/api/attachments/198"}],
             "Title": "Baugesetz"
+        }, {
+            "Canton": "BL",
+            "DocumentType": "LegalProvision",
+            "Lawstatus_Code": "inKraft",
+            "Lawstatus_Text": "Rechtskräftig",
+            "OfficialNumber": "3891.100",
```

Comment:
@michmuel 
can there be 2 records with same title but different OfficialNumber ?
In that case you could add one more record into this test and check that OfficialNumbers are sorted correctly (as a third rank sort criteria after title and number-not-empty)

Reply:
@mki-c2c In BL we have legal provisions with same title from which only one has a number. I think there are no restrictions concerning the tittle-number combinations in the official OEREB docs.

I split the tests in Baugesetz2 and Baugesetz3 as the sort function is not sensitive to official number not defined and official number=empty string.

Reply:
Why is the sorting done on Title and OfficialNumber? Should we not use the index attribute here to sort the legal_provisions (and the laws and the hints)? In the federal spec for the pdf there is a certain order required (federal laws => federal ordinances => cantonal laws => cantonal ordinances). This order cannot be produced by sorting on title and number. In the JSON extract the order is as required.

Reply:
I agree on that. Sorting by index would be the most clean and preferable solution. But @jwkaltz should have the best insight what is requested by the members so far.

@michmuel Did you checked if you can use sort by index in your case to get the output you want?

Reply:
I can create a branch with that capability and test it in our environment.

Reply:
@jwkaltz Where do you think the dcuments are sorted? I did not found it in https://github.com/openoereb/pyramid_oereb/blob/b8620f0372c6951b150c5bb999f17d7b3ce193ca/pyramid_oereb/core/renderer/extract/json_.py

I guess sorting in mapfish_print.py is fine as "Index" is mandatory and is the parameter for this purpose.

Reply:
If we sort them already when preparing the data extract, it would need to be at the place where we gather the documents, so that it's already sorted when serialized into JSON or XML.
Alternatively, each consumer needs to sort them, thus mapfish_print.py, and all oereb client implementations.

Reply:
I completely agree-we should sort the documents in the core already. Like we do for themes. In the moment this is not done or implemented in pyramid_oereb

Reply:
I am not sure about that as there is no specification that elements in the extracts have to be sorted. Only the information is mandatory that enables sorting.

Reply:
hmm this is true too

Reply:
This is, why the index is delivered in the extract json/xml

Reply:
I created a PR to test sorting by index: https://github.com/openoereb/pyramid_oereb/pull/1504. Some comments are added there.
A general thing concerning the tests (@mki-c2c, @vvmruder): the json at https://github.com/openoereb/pyramid_oereb/blob/bugfix-1474/tests/contrib.print_proxy.mapfish_print/resources/test_extract.json still is in the old schema version 1.0.

Reply:
sorry for my late answer, I was really busy

Sorting of documents is defined in appendix E, page 50 and following:
`/** Ordnungszahl für die Sortierung im Auszug **/
AuszugIndex : MANDATORY -1000 .. 1000;`

A lot of the positive range is already blocked by the way the confederation defines it's indexes for the legal bases, e.g.:
> 000 Raumplanung
010 Bundesgesetz über die Raumplanung (SR 700)
100 Strassen
110 Bundesgesetz über die Nationalstrassen (SR 725.11)
120 Nationalstrassenverordnung (SR 725.111)

leaving following ranges for legal bases on state or municipality level:

> 000 Raumplanung
020 bis und mit 099 für kantonale und kommunale gesetzliche Grundlagen der Kategorie Raumplanung
100 Strassen
130 bis und mit 199 für kantonale und kommunale gesetzliche Grundlagen der Kategorie Strassen

In the Extract-model, the Index is also defined. But we can not use only the index to sort the documents, we need a second criteria - either the document type or alphabetical order on the title or something else.

Why? - It may happen that you have more legal bases on cantonal level than the available index range for a theme (but unlikely)
But we need to sort the legal dispositions too and there obviously exist more documents for some themes than the available ranges.
So, I'd suggest we use the document type & index as the sort criteria which allows us to use the complete range from -1000 to 1000 for other documents than legal bases. - I'm still unsure wheter 2000 positions are sufficient for large cantons in the Land Use Plans theme...

Reply:
In https://github.com/openoereb/pyramid_oereb/pull/1504 sorting is by index separately performed for 'LegalProvisions', 'Laws', 'Hints'.

It is possible that the index range is to small to be able to assign a number to each cantonal legal provision. However, the index does not have to be unique. Two documents that are never related to the same PLR may have the same index.

Reply:
Discussion in the pyramid_oereb weekly on 2022-02-14. Decision that sorting should be performed by index separately in the categories "legal_provisions" / "laws" / "hints". This PR is closed and the issue is continued in https://github.com/openoereb/pyramid_oereb/pull/1504.